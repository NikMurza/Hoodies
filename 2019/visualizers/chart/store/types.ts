import { ChartData } from '@common/api/types/chart';

export interface ChartVisualizerInstance {
  openId: string;
  data?: ChartData;
  isFetching: boolean;
}

export interface ChartVisualizersState {
  [instanceId: string]: ChartVisualizerInstance;
}

export interface StatePart {
  chartVisualizers?: ChartVisualizersState;
}
export type GetState = () => StatePart;

export interface ChartPayload<T> {
  instanceId: string;
  data?: T;
}
