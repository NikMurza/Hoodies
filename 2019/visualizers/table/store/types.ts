import { PPSom } from '@common/api/types/PPSom';

export interface TableVisualizerInstance {
  openId: string;
  tabSheetData: TabSheetRawData;
  isFetching: boolean;
}

export interface TableVisualizersState {
  [instanceId: string]: TableVisualizerInstance;
}

export interface StatePart {
  tableVisualizers?: TableVisualizersState;
}
export type GetState = () => StatePart;

export interface TabSheetPayload<T> {
  instanceId: string;
  data?: T;
}

export interface TabSheetRawData {
  /** Структура и диапазоны данных. */
  Data?: PPSom.TabData[];
  /** Загруженные диапазоны. */
  Ranges?: PPSom.TabRange[];
  /** Метаданные */
  Metadata?: PPSom.TabMd;
  /** Выделение (Дублируется из метаданных). */
  Selection?: PPSom.TabSelection;
}
