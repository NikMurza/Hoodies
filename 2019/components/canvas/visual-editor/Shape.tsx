import * as React from 'react';
import { mapTypeToComponent } from '@common/components/lib/canvas/shapes';
import {
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorShape,
  VisualEditorSize,
  VisualEditorComponentData,
} from '@common/components/lib/types/visual-editor';
import { typeContextMenuId } from '@common/components/lib/canvas/components/context-menu/with-context-menu/WithContextMenu';

interface Props {
  item: VisualEditorShape;
  contextMenuId: string;
  stageSize: VisualEditorSize;
  hoveredShape: { id: string; type: string };
  onShapeAction: <T extends VisualEditorContextMenuIdentity>(
    action: VisualEditorActions,
    data: T,
  ) => void;
  onShapeAddComponent: (data: VisualEditorComponentData) => void;
  onShapeRemoveComponent: (id: string) => void;
}

const Shape: React.FC<Props> = ({
  item,
  contextMenuId,
  stageSize,
  hoveredShape,
  onShapeAction,
  onShapeAddComponent,
  onShapeRemoveComponent,
}) => {
  const { id, type } = item;
  const ShapeComponent = mapTypeToComponent[type];
  const isContextMenuShow = contextMenuId === typeContextMenuId(id, type);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const props = item as any;

  // перехватываем onAction у shape
  const onAction = React.useCallback(
    <T extends VisualEditorContextMenuIdentity>(action: VisualEditorActions, data: T): void => {
      onShapeAction(action, data);
      if (props.onAction) props.onAction(action, data);
    },
    [props, onShapeAction],
  );
  return (
    <ShapeComponent
      key={`${type}_${id}`}
      {...props}
      onAction={onAction}
      isHovered={id === hoveredShape.id && type === hoveredShape.type}
      isContextMenuShow={isContextMenuShow}
      stageSize={stageSize}
      onAddComponent={onShapeAddComponent}
      onRemoveComponent={onShapeRemoveComponent}
    />
  );
};

export default Shape;
