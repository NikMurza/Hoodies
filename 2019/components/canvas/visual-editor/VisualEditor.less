.visual-editor {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;

  & .visual-canvas {
    position: relative;
    overflow: hidden;
    outline: none;

    &-scrollable {
      position: relative;
      flex-grow: 1;
      overflow: auto;
      &_hidden-scroll {
        -ms-overflow-style: none;
        scrollbar-width: none;
      }
      &_hidden-scroll::-webkit-scrollbar {
        display: none;
      }
    }
    &__components {
      position: absolute;
      top: 0;
      left: 0;
    }
    &__context-menu-wrapper {
      position: absolute;
    }
  }
}
