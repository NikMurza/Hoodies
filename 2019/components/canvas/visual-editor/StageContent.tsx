import * as React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import * as ReactKonva from 'react-konva';

import {
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorShape,
  VisualEditorSize,
  VisualEditorComponentData,
} from '@common/components/lib/types/visual-editor';
import Shape from './Shape';

export interface Props {
  store?: Store;
  items?: VisualEditorShape[];
  hoveredShape: { id: string; type: string };
  contextMenuId: string;
  stageSize: VisualEditorSize;
  onShapeAction: <T extends VisualEditorContextMenuIdentity>(
    action: VisualEditorActions,
    data: T,
  ) => void;
  onShapeAddComponent: (data: VisualEditorComponentData) => void;
  onShapeRemoveComponent: (id: string) => void;
}

const StageContent: React.FC<Props> = ({
  store,
  items,
  hoveredShape,
  contextMenuId,
  stageSize,
  onShapeAction,
  onShapeAddComponent,
  onShapeRemoveComponent,
}) => {
  const stage = React.useMemo(() => {
    const isHoveredShape = (shape: VisualEditorShape): boolean =>
      shape.id === hoveredShape.id && shape.type === hoveredShape.type;
    const ordered = items.filter((item) => !isHoveredShape(item));
    ordered.push(...items.filter(isHoveredShape));

    const renderItem = (item: VisualEditorShape): JSX.Element =>
      React.createElement(Shape, {
        key: `${item.type}-${item.id}`,
        item,
        contextMenuId,
        stageSize,
        hoveredShape,
        onShapeAction,
        onShapeAddComponent,
        onShapeRemoveComponent,
      });

    return <ReactKonva.Layer>{ordered.map(renderItem)}</ReactKonva.Layer>;
  }, [
    items,
    hoveredShape,
    contextMenuId,
    stageSize,
    onShapeAction,
    onShapeAddComponent,
    onShapeRemoveComponent,
  ]);

  return store ? <Provider store={store}>{stage}</Provider> : stage;
};

export default StageContent;
