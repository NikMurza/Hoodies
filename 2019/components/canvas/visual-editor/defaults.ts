import colorsConst from '@common/styles/constants/colors';
import fontConst from '@common/styles/constants/fonts';
import { VisualEditorSize } from '@common/components/lib/types/visual-editor';

interface ScaleDefaults {
  initial: number;
  min: number;
  max: number;
  step: number;
}
export interface CanvasDefaults {
  size: VisualEditorSize;
  gridCell: number;
  gridLineWidth: number;
  canvasOffset: number;
  zoom: ScaleDefaults;
}

// измерения холста
export const canvas: CanvasDefaults = {
  size: {
    width: window.innerWidth,
    height: window.innerHeight,
  },
  gridCell: 96, // сторона ячейки сетки
  gridLineWidth: 1,
  canvasOffset: 24, // минимально допустимое расстояние от края холста до элемента,
  zoom: {
    initial: 1,
    min: 0.1,
    max: 4,
    step: 0.1,
  },
};

// размер элемента по-умолчанию
export const element = {
  width: 200,
  height: 100,
  indent: 100, // размер отступа (по-умолчанию) между элементами
  radius: 60, // Радиус элемента (по-умолчанию)
  borderRadius: 0,
  borderWidth: 1,
  referencePointRadius: 6,
  editInputWidth: 100,
  editInputHeight: 24,
};

export const line = {
  borderWidth: 2,
  hoverWidth: 20,
  endpointLength: 11.5,
  endpointWidth: 10,
  iconRadius: 16,
};

export const font = {
  family: fontConst.family,
  textSize: 14,
  accentTextSize: 16,
  secondaryTextSize: 13,
};

export const colors = {
  backgroundColor: colorsConst.white,
  transparent: 'transparent',
  gridLinesColor: '#ECEDEF',
  shadowColor: '#cccccc',
  canvas: '#F2F3F4',
  circleColor: '#9DAFED',
  circleBackgroundColor: '#A469AD',
  circleBorderColor: '#8B5A94',
  circleActiveBackgroundColor: '#3B5EDC',
  circleActiveBorderColor: '#FFFFFF',
  subTextColor: '#93989E',
  sourceTypeColor: colorsConst.green,
  destTypeColor: '#59BDF8',
  convertorTypeColor: '#FFC700',
  blockShadowColor: 'rgba(115,115,115,0.12)',
  blockActiveColor: colorsConst.blue,
  borderColor: colorsConst.darkAdditional,
  referencePointsColor: colorsConst.white,
  referencePointsHoverColor: colorsConst.blue,
  referencePointsBorderColor: colorsConst.blue,
  iconSettingBackgroundColor: colorsConst.white,
  iconSettingColor: colorsConst.white,
};
