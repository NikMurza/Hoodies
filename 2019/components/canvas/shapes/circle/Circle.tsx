import * as React from 'react';
import * as ReactKonva from 'react-konva';
import Konva from 'konva';
// eslint-disable-next-line import/no-unresolved
import { KonvaEventObject } from 'konva/types/Node';
import {
  VisualEditorCircle,
  VisualEditorShapeType,
  VisualEditorComponentData,
  VisualEditorActions,
  VisualEditorComponentContract,
  ContextMenuItem,
  VisualEditorPosition,
  VisualEditorActionClick,
} from '@common/components/lib/types/visual-editor';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import * as color from '@common/styles/constants/colors';
import onMouseOverCursorPointer from '@common/hocs/lib/onMouseOverCursorPointer/onMouseOverCursorPointer';
import Input from '@common/components/lib/canvas/components/input/input';
import WithContextMenu from '@common/components/lib/canvas/components/context-menu/with-context-menu/WithContextMenu';
import WithHover from '@common/components/lib/canvas/components/with-hover/WithHover';
import WithDrag from '@common/components/lib/canvas/components/with-drag/WithDrag';
import { useExtension } from '../../hooks/useExtension';

const { canvasOffset } = defaults.canvas;

export interface CircleProps extends VisualEditorCircle, VisualEditorComponentContract {
  contextMenuItems?: ContextMenuItem[];
  isContextMenuShow?: boolean;
  onAction?<T extends object>(action: VisualEditorActions, data: T): void;
  onAddComponent?: (data: VisualEditorComponentData) => void;
  onRemoveComponent?: (id: string) => void;
  isHovered?: boolean;
}

// Группа с событиями на mouse over - изменение курсора
const GroupWithHOCs = WithHover(
  WithDrag(
    WithContextMenu(
      onMouseOverCursorPointer((props: Konva.NodeConfig) => <ReactKonva.Group {...props} />),
    ),
  ),
);

// расширить область hover
const extendHoverArea = 2;

/**
 * Отрисовывает Круг с текстом
 * @param props Свойства
 */
const Circle: React.FC<CircleProps> = (props: CircleProps): JSX.Element => {
  const {
    id,
    type,
    text,
    textColor,
    // х и у отвечают за центр круга.
    x,
    y,
    stageSize,
    radius,
    backgroundColor,
    borderColor,
    contextMenuItems,
    isContextMenuShow,
    onAction,
    onAddComponent,
    onRemoveComponent,
    extension,
  } = props;
  const inputId = `input_${id}`;

  const [isEditing, setIsEditing] = React.useState(false);

  const enableEditing = React.useCallback((): void => setIsEditing(true), []);

  const onComplete = React.useCallback(
    (newText: string): void => {
      if (newText !== text) {
        onAction<VisualEditorCircle>(VisualEditorActions.Update, {
          id,
          type,
          x,
          y,
          radius,
          backgroundColor,
          text: newText,
        });
      }
      setIsEditing(false);
    },
    [type, text, id, x, y, radius, backgroundColor, onAction],
  );

  const onCancel = React.useCallback((): void => {
    setIsEditing(false);
  }, []);

  React.useEffect(() => {
    if (isEditing) {
      const style = {
        color: textColor,
      };

      const data = {
        id: inputId,
        component: (
          <Input
            key={inputId}
            id={inputId}
            text={text}
            x={x}
            y={y}
            onComplete={onComplete}
            onCancel={onCancel}
            style={style}
          />
        ),
      };
      onAddComponent(data);
    } else {
      onRemoveComponent(inputId);
    }
  }, [isEditing, textColor, text, x, y, onComplete, onCancel]);

  const onGroupAction = React.useCallback(
    <T extends object>(action: VisualEditorActions, data: T): void => {
      const update = (): void => {
        // дополняем передаваемые данные до полного интерфейса VisualEditorCircle
        data = { ...data, type, radius, backgroundColor, text };
      };

      const actions = {
        [VisualEditorActions.Edit]: enableEditing,
        [VisualEditorActions.Update]: update,
        [VisualEditorActions.Move]: update,
      };

      // вызываем обработчики
      if (actions[action]) actions[action]();

      // прокидываем наверх
      onAction(action, data);
    },
    [type, radius, backgroundColor, text, onAction],
  );

  const onActionHandler = React.useCallback(
    (action: VisualEditorActions) => (event: KonvaEventObject<MouseEvent>): void => {
      onAction<VisualEditorActionClick>(action, {
        id,
        event: event?.evt,
      });
    },
    [id, onAction],
  );

  // не даем унести элемент за пределы холста (граница - canvasOffset от края)
  const getRestrictedCoordinate = React.useCallback(
    (coord: number, canvasEdgeCoord: number): number => {
      let newCoord: number;

      // минимально и максимально допустимые координаты центра circle
      const minCoord = radius + canvasOffset;
      const maxCoord = canvasEdgeCoord - radius - canvasOffset;

      if (coord < minCoord) {
        newCoord = minCoord;
      } else if (coord > maxCoord) {
        newCoord = maxCoord;
      } else {
        newCoord = coord;
      }
      return newCoord;
    },
    [radius],
  );

  const dragBoundFunc = React.useCallback(
    (pos: VisualEditorPosition): VisualEditorPosition => ({
      x: getRestrictedCoordinate(pos.x, stageSize.width),
      y: getRestrictedCoordinate(pos.y, stageSize.height),
    }),
    [stageSize],
  );

  const extensionElement = useExtension(extension, props);

  return (
    <GroupWithHOCs
      id={id}
      type={type}
      x={x}
      y={y}
      isContextMenuShow={isContextMenuShow}
      items={contextMenuItems}
      onAction={onGroupAction}
      onAddComponent={onAddComponent}
      onRemoveComponent={onRemoveComponent}
      draggable={!isEditing}
      stageSize={stageSize}
      dragBoundFunc={dragBoundFunc}
      onDblClick={enableEditing}
      onClick={onActionHandler(VisualEditorActions.Click)}
      onMouseUp={onActionHandler(VisualEditorActions.MouseUp)}
    >
      {/* рисуем невидимый круг - чтобы расширить область hover у группы */}
      <ReactKonva.Circle
        radius={radius + defaults.element.referencePointRadius + extendHoverArea}
      />
      {/* сам компонент */}
      <ReactKonva.Group>
        <ReactKonva.Circle
          id={id}
          radius={radius}
          fill={backgroundColor}
          stroke={borderColor}
          shadowColor={defaults.colors.shadowColor}
          shadowOffsetY={2}
          shadowBlur={4}
        />
        {/* Не показываем текст в режиме редактирования */}
        {!isEditing ? (
          <ReactKonva.Text
            // Сместить текст к центру координат на 0.75 радиуса
            x={-0.758 * radius}
            // Поднять текст на 50% радиуса
            y={-0.451 * defaults.font.textSize}
            fill={textColor}
            text={`${text}`}
            width={1.5 * radius}
            fontSize={defaults.font.textSize}
            wrap="none"
            align="center"
            verticalAlign="middle"
            ellipsis
            fontFamily={defaults.font.family}
          />
        ) : null}
      </ReactKonva.Group>

      {/* отрисовываем extension */}
      {extensionElement}
    </GroupWithHOCs>
  );
};

interface CreateCircleOpts {
  id?: string;
  text?: string;
  x?: number;
  y?: number;
  radius?: number;
  textColor?: string;
  backgroundColor?: string;
  borderColor?: string;
  contextMenuItems?: ContextMenuItem[];
  onAction?: <T extends object>(action: VisualEditorActions, data: T) => void;
  isReferencePointsShow?: boolean;
  extension?: React.ComponentType;
}

/**
 * Создание элемента "Круг с текстом" со всеми параметрами
 * @param {CreateCircleOpts} opts Опции
 */
export const createCircle = (opts?: CreateCircleOpts): VisualEditorCircle => {
  opts = opts || {};
  const emptyFn = (): void => undefined;

  return {
    type: VisualEditorShapeType.Circle,
    id: opts.id || '',
    extension: opts.extension || null,
    text: opts.text || '',
    backgroundColor: opts.backgroundColor || defaults.colors.circleBackgroundColor,
    borderColor: opts.borderColor || defaults.colors.circleBorderColor,
    textColor: opts.textColor || color.default.white,
    radius: opts.radius || defaults.element.radius,
    x: opts.x,
    y: opts.y,
    contextMenuItems: opts.contextMenuItems || [],
    onAction: opts.onAction || emptyFn,
  };
};

Circle.defaultProps = {
  ...createCircle(),
  onAddComponent: () => (): void => undefined,
  onRemoveComponent: () => (): void => undefined,
};

export default Circle;
