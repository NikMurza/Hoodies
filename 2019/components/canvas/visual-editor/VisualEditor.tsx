import * as React from 'react';

import { ReactReduxContext } from 'react-redux';
import Konva from 'konva';
import * as ReactKonva from 'react-konva';

import './VisualEditor.less';
import {
  VisualEditorActionCallback,
  VisualEditorActions,
  VisualEditorContextMenuIdentity,
  VisualEditorIdentityPosition,
  VisualEditorKeyboardAction,
  VisualEditorPosition,
  VisualEditorPositionalClick,
  VisualEditorShape,
  VisualEditorShapeWithRef,
  VisualEditorSize,
} from '@common/components/lib/types/visual-editor';
import { MouseButtons } from '@common/components/lib/types/mouseEvent';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import { usePosition } from '@common/components/lib/canvas/hooks/usePosition';
import { useIsStage } from '@common/components/lib/canvas/hooks/useIsStage';
import WithDrop from '@common/components/lib/canvas/visual-editor/with-drop/WithDrop';
import WithResizeDetect, {
  WithResizeable,
} from '@common/components/lib/canvas/visual-editor/with-resize-detect/WithResizeDetect';
import StageContent from './StageContent';
import useAddRemoveComponents from './use-add-remove-component/useAddRemoveComponent';

export interface VELayer extends Konva.LayerConfig {
  children: JSX.Element;
}

export interface Props {
  /**
   * Идентификатор конкретного экземпляра VE
   */
  id?: string;

  /**
   * Элементы для отрисовки
   */
  items?: VisualEditorShape[];

  /*
   * Дополнительные слои ниже основного с контентом
   * Рекомендуется в них отображать дополнительные данные — grid, FPS и пр
   * См. useFPS, useGrid, useZoom
   */
  extraLayers?: JSX.Element;

  /*
   * Положение Stage
   * Значение по умолчанию - { x: 0, y: 0 }
   */
  position?: VisualEditorPosition;

  /*
   * Масштабирование содержимого Visual Editor, указывается в долях от 0,1 и выше.
   * Значение по умолчанию - 1, что соответствует 100%
   */
  zoom?: number;

  /*
   * Контролируем возможность перемещаться по канвасу
   * По умолчанию true
   */
  panEnabled?: boolean;

  /**
   * Вызывается на каких-либо событиях (изменение, удаление, контекстное меню)
   * @param action Событие
   * @param data Данные
   */
  onAction?: VisualEditorActionCallback;
}

const VisualEditorContainer = WithDrop(
  WithResizeDetect(
    (
      props: React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement> &
        WithResizeable,
    ): JSX.Element => {
      const { resizeableRef, ...rest } = props;
      return <div ref={resizeableRef} {...rest} />;
    },
  ),
);

/**
 * Визуальный редактор - отображает таблицы со связями (и не только)
 * @param props Свойства
 */
const VisualEditor: React.FC<React.PropsWithChildren<Props>> = ({
  id,
  position,
  zoom: customZoom = 1,
  items,
  extraLayers,
  children,
  panEnabled,
  onAction,
}) => {
  const [contextMenuId, setContextMenuId] = React.useState<string>('');
  const [hoveredShape, setHoveredShape] = React.useState({ id: '', type: '' });

  const [stageSize, setStageSize] = React.useState<VisualEditorSize>({
    width: window.innerWidth,
    height: window.innerHeight,
  });

  const stagePosition: VisualEditorPosition = {
    x: position?.x ?? 0,
    y: position?.y ?? 0,
  };

  const zoom = customZoom ?? defaults.canvas.zoom.initial;

  const stageRef = React.useRef<ReactKonva.Stage>(null);
  const isStage = useIsStage<Event>();

  const { components, onShapeAddComponent, onShapeRemoveComponent } = useAddRemoveComponents();

  const hideContextMenu = React.useCallback((): void => {
    setContextMenuId('');
  }, []);

  React.useEffect(() => {
    onAction<VisualEditorShapeWithRef<ReactKonva.Stage>>(VisualEditorActions.Mount, {
      forwardRef: stageRef,
    });
  }, [stageRef]);

  // Закрываем контекстное меню при правом клике на пустую область,
  // восстанавливаем действие по умолчанию.
  const onStageContextMenu = React.useCallback(
    (e: Konva.KonvaEventObject<PointerEvent>): void => {
      if (isStage(e)) {
        e.evt.preventDefault();
        const { x, y } = usePosition<PointerEvent>(e);

        onAction<VisualEditorPositionalClick>(VisualEditorActions.ContextMenu, {
          id,
          x,
          y,
          target: e.target,
          event: { ...e?.evt, button: MouseButtons.Right },
        });
      }
    },
    [hideContextMenu, onAction],
  );

  const onStageClick = React.useCallback(
    (e: Konva.KonvaEventObject<PointerEvent>): void => {
      const { x, y } = usePosition<PointerEvent>(e);
      onAction<VisualEditorPositionalClick>(VisualEditorActions.Click, {
        id,
        x,
        y,
        event: e?.evt,
        target: e?.target,
      });
      hideContextMenu();
    },
    [id, onAction],
  );

  const onStageMouseAction = React.useCallback(
    (action: VisualEditorActions) => (e: Konva.KonvaEventObject<MouseEvent>): void => {
      const { x, y } = usePosition<MouseEvent>(e);
      onAction<VisualEditorIdentityPosition>(action, {
        id,
        x,
        y,
      });
    },
    [id, onAction],
  );

  const onStageMouseMove = React.useCallback(onStageMouseAction(VisualEditorActions.MouseMove), [
    stageRef,
    onStageMouseAction,
  ]);

  const onStageMouseUp = React.useCallback(onStageMouseAction(VisualEditorActions.MouseUp), [
    stageRef,
    onStageMouseAction,
  ]);

  const onKeyboardAction = React.useCallback(
    (key: VisualEditorActions, event: React.KeyboardEvent<HTMLDivElement>): void => {
      onAction<VisualEditorKeyboardAction>(key, {
        id,
        isCtrl: event.ctrlKey,
        isShift: event.shiftKey,
        isAltKey: event.altKey,
        keyCode: event.keyCode,
        key: event.key,
        isRepeat: event.repeat,
      });
    },
    [id, onAction],
  );

  const onStageKeyDown = React.useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>): void =>
      onKeyboardAction(VisualEditorActions.KeyDown, event),
    [onKeyboardAction],
  );

  const onStageKeyUp = React.useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>): void =>
      onKeyboardAction(VisualEditorActions.KeyUp, event),
    [onKeyboardAction],
  );

  const onStageDrag = React.useCallback(
    (e: Konva.KonvaEventObject<DragEvent>) => {
      if (!isStage(e)) {
        return;
      }

      onAction<VisualEditorIdentityPosition>(VisualEditorActions.Move, {
        id,
        x: -e.target.x(),
        y: -e.target.y(),
      });
    },
    [id, onAction],
  );

  const onStageResize = React.useCallback(
    (size: VisualEditorSize): void => {
      setStageSize(size);
      onAction<VisualEditorSize>(VisualEditorActions.Resize, size);
    },
    [onAction],
  );

  const onShapeAction = React.useCallback(
    <T extends VisualEditorContextMenuIdentity>(action: VisualEditorActions, data: T): void => {
      const actions = {
        [VisualEditorActions.ContextMenu]: (data: VisualEditorContextMenuIdentity): void =>
          setContextMenuId(data.contextMenuId),
        [VisualEditorActions.Move]: hideContextMenu,
        [VisualEditorActions.Hover]: (data: VisualEditorShape): void => {
          setHoveredShape({ id: data.id, type: data.type });
        },
      };

      if (actions[action]) actions[action](data);
    },
    [hideContextMenu],
  );

  React.useEffect(() => {
    const selector = document.getElementById('root');
    if (!selector) {
      return null;
    }
    selector.addEventListener('click', hideContextMenu);
    selector.addEventListener('contextmenu', hideContextMenu, true);
    return (): void => {
      selector.removeEventListener('click', hideContextMenu);
      selector.removeEventListener('contextmenu', hideContextMenu);
    };
  }, []);

  const renderStage: boolean = stageSize.width > 0 && stageSize.height > 0;

  return (
    <VisualEditorContainer
      onResize={onStageResize}
      className="visual-editor"
      stageRef={stageRef}
      onAction={onAction}
    >
      <div
        tabIndex={-1}
        className="visual-canvas"
        onKeyDown={onStageKeyDown}
        onKeyUp={onStageKeyUp}
      >
        <ReactReduxContext.Consumer>
          {({ store }): JSX.Element =>
            !renderStage ? null : (
              <ReactKonva.Stage
                draggable={panEnabled}
                ref={stageRef}
                id={id}
                x={-stagePosition.x}
                y={-stagePosition.y}
                width={stageSize.width}
                height={stageSize.height}
                scaleX={zoom}
                scaleY={zoom}
                stageRef={stageRef}
                onClick={onStageClick}
                onContextMenu={onStageContextMenu}
                onMouseMove={onStageMouseMove}
                onMouseUp={onStageMouseUp}
                onDragEnd={onStageDrag}
              >
                {extraLayers}
                {/* подключаем к стору конву, если задан store */}
                {React.createElement(StageContent, {
                  store,
                  items,
                  hoveredShape,
                  contextMenuId,
                  stageSize,
                  onShapeAction,
                  onShapeAddComponent,
                  onShapeRemoveComponent,
                })}
              </ReactKonva.Stage>
            )
          }
        </ReactReduxContext.Consumer>
        {/* отрисовываем дополнительные компоненты */}
        <div className="visual-canvas__components">
          {Object.values(components)}
          {/* отрисовываем детей */}
          {children}
        </div>
      </div>
    </VisualEditorContainer>
  );
};

// по-умолчанию редактор занимает размер экрана
VisualEditor.defaultProps = {
  id: 'VisualEditor',
  items: [],
  panEnabled: true,
  onAction: (): void => undefined,
};

export default VisualEditor;
