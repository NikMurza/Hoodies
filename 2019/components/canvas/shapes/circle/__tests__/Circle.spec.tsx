import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import * as defaults from '@common/components/lib/canvas/visual-editor/defaults';
import * as color from '@common/styles/constants/colors';
import Circle, { createCircle } from '../Circle';
import { VisualEditorShapeType, VisualEditorActions } from '../../../../types/visual-editor';

describe('Circle component', () => {
  it('renders correctly', () => {
    const type = VisualEditorShapeType.Circle;
    const id = 'table1';
    const x = 100;
    const y = 100;
    const radius = 100;

    const props = {
      onAction: jest.fn,
    };

    const onActionSpy = jest.spyOn(props, 'onAction');

    const wrapper = shallow(
      <Circle id={id} radius={radius} type={type} x={x} y={y} onAction={props.onAction} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.simulate('action');
    expect(onActionSpy).toBeCalled();
  });

  it('createCircle works correctly', () => {
    const id = 'table1';
    expect(createCircle({ id, text: id, onAction: jest.fn })).toEqual({
      type: VisualEditorShapeType.Circle,
      id,
      text: id,
      backgroundColor: defaults.colors.circleBackgroundColor,
      borderColor: defaults.colors.circleBorderColor,
      textColor: color.default.white,
      radius: defaults.element.radius,
      x: undefined,
      y: undefined,
      contextMenuItems: [],
      onAction: jest.fn,
      extension: null,
    });
  });

  it('Circle defaultProps', () => {
    expect(JSON.stringify(Circle.defaultProps)).toEqual(
      JSON.stringify({
        ...createCircle(),
        onAddComponent: () => (): void => undefined,
        onRemoveComponent: () => (): void => undefined,
      }),
    );

    expect(Circle.defaultProps.onAction(VisualEditorActions.Add, {})).toEqual(undefined);
    expect(JSON.stringify(Circle.defaultProps.onAddComponent)).toEqual(undefined);
    expect(JSON.stringify(Circle.defaultProps.onRemoveComponent)).toEqual(undefined);
  });

  it('createCircle ext FC works correctly', () => {
    const id = 'table1';
    const Ext1: React.FC = (): JSX.Element => {
      return <div>test extension 1</div>;
    };

    expect(createCircle({ id, text: id, onAction: jest.fn, extension: Ext1 })).toEqual({
      type: VisualEditorShapeType.Circle,
      id,
      text: id,
      backgroundColor: defaults.colors.circleBackgroundColor,
      borderColor: defaults.colors.circleBorderColor,
      textColor: color.default.white,
      radius: defaults.element.radius,
      x: undefined,
      y: undefined,
      contextMenuItems: [],
      onAction: jest.fn,
      extension: Ext1,
    });
  });

  it('createCircle ext Component works correctly', () => {
    const id = 'table1';
    class Ext2 extends React.Component {
      render() {
        return <div>test extension 2</div>;
      }
    }

    expect(createCircle({ id, text: id, onAction: jest.fn, extension: Ext2 })).toEqual({
      type: VisualEditorShapeType.Circle,
      id,
      text: id,
      backgroundColor: defaults.colors.circleBackgroundColor,
      borderColor: defaults.colors.circleBorderColor,
      textColor: color.default.white,
      radius: defaults.element.radius,
      x: undefined,
      y: undefined,
      contextMenuItems: [],
      onAction: jest.fn,
      extension: Ext2,
    });
  });
});
