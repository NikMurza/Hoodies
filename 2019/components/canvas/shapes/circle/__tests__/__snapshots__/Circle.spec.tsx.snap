// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle component renders correctly 1`] = `
<Component
  dragBoundFunc={[Function]}
  draggable={true}
  id="table1"
  items={Array []}
  onAction={[Function]}
  onAddComponent={[Function]}
  onClick={[Function]}
  onDblClick={[Function]}
  onMouseUp={[Function]}
  onRemoveComponent={[Function]}
  type="circle"
  x={100}
  y={100}
>
  <Circle
    radius={108}
  />
  <Group>
    <Circle
      fill="#A469AD"
      id="table1"
      radius={100}
      shadowBlur={4}
      shadowColor="#cccccc"
      shadowOffsetY={2}
      stroke="#8B5A94"
    />
    <Text
      align="center"
      ellipsis={true}
      fill="#FFF"
      fontFamily="OpenSans, sans-serif"
      fontSize={14}
      text=""
      verticalAlign="middle"
      width={150}
      wrap="none"
      x={-75.8}
      y={-6.314}
    />
  </Group>
</Component>
`;
